### ISSUE TITLE
MRFC 5 - Concatenate multiples binaries into a static binary à-la-busybox

### ISSUE THREAD

**Issue by moul** [moul/ideas#5]
i.e: concatenate `/bin/bash` and `/bin/tar` into a unique binary with smart entrypoint based on `argv[0]`, so we can generate symlinks à-la-busybox

goal: create a busybox-like binary with full-featured **tar**, full-featured **wget** etc

**Comment by aimxhaisse** [comment]
Random approach to this cool problem:
- [ ] compress all binaries and embed them in char \* (gzip or whatever)
- [ ] in the 'master' busybox: unzip in memory the binary (from the char *), mmap it to a file, fork + exec that file, and profit
- [ ] profit again

**Comment by bchatelard** [comment]
Bonus: Handle multi arch binaries

**Comment by aimxhaisse** [comment]
Bonus point: minimizer (i.e: a big static binary such as firefox, could be nicely compressed and packed)

**Comment by aimxhaisse** [comment]
Bonus point: embed assets in the binary, and provide them on demand to the process ran: ptrace and upon open() syscalls, check if this is an asset we have in the binary, if so, provide it

**Comment by moul** [comment]
Bonus: generate `FROM scratch; ADD bin; ENTRYPOINT ["bin"]` minimalist Docker images

**Comment by aimxhaisse** [comment]
Bonus point 2000: an on-demand Linux distribution, download remote binaries on-demand and map them

**Comment by aimxhaisse** [comment]
Bonus: create a binary that spawns a docker server with non-conflicting params (weird port, etc), that way we could embed a docker image in it, and run it on the fly, basically create a all-in-one way to run a docker instance.

**Comment by moul** [comment]
Yes, but instead of Docker, we can embed [runc](https://github.com/opencontainers/runc) + the image

**Comment by aimxhaisse** [comment]
:ok_hand:

**Comment by moul** [comment]
Here is a working POC https://github.com/moul/mister-bin
Now I need to add helpers for libraries&assets + document some use cases
